### DISCUSSION TITLE
FCStd more info

### DISCUSSION THREAD

**Discussion by zeffii** [nortikin/sverchok#4526] [Ideas]
@yorikvanhavre wrote an importer for [FCStd files a few moons ago](https://gist.github.com/yorikvanhavre/680156f59e2b42df8f5f5391cae2660b). In the spirit of open access to data and tools , here's an initial port of that code to Sverchok snlite script node.

now outdated code here..

<details>

```python
"""
>in FilePath FP
out names T
out diffuse_colors v
"""
import zipfile
import xml.sax
import FreeCAD

class FreeCAD_xml_handler(xml.sax.ContentHandler):

    """A XML handler to process the FreeCAD GUI xml data"""

    # this creates a dictionary where each key is a FC object name,
    # and each value is a dictionary of property:value pairs

    def __init__(self):

        self.guidata = {}
        self.current = None
        self.properties = {}
        self.currentprop = None
        self.currentval = None

    # Call when an element starts

    def startElement(self, tag, attributes):

        if tag == "ViewProvider":
            self.current = attributes["name"]
        elif tag == "Property":
            name = attributes["name"]
            if name in ["Visibility","ShapeColor","Transparency","DiffuseColor"]:
                self.currentprop = name
        elif tag == "Bool":
            if attributes["value"] == "true":
                self.currentval = True
            else:
                self.currentval = False
        elif tag == "PropertyColor":
            c = int(attributes["value"])
            r = float((c>>24)&0xFF)/255.0
            g = float((c>>16)&0xFF)/255.0
            b = float((c>>8)&0xFF)/255.0
            self.currentval = (r,g,b)
        elif tag == "Integer":
            self.currentval = int(attributes["value"])
        elif tag == "Float":
            self.currentval = float(attributes["value"])
        elif tag == "ColorList":
            self.currentval = attributes["file"]

    # Call when an elements ends

    def endElement(self, tag):

        if tag == "ViewProvider":
            if self.current and self.properties:
                self.guidata[self.current] = self.properties
                self.current = None
                self.properties = {}
        elif tag == "Property":
            if self.currentprop and (self.currentval != None):
                self.properties[self.currentprop] = self.currentval
                self.currentprop = None
                self.currentval = None

filename = FilePath[0][0]
# check if we have a GUI document
guidata = {}
zdoc = zipfile.ZipFile(filename)
if zdoc:
    if "GuiDocument.xml" in zdoc.namelist():
        gf = zdoc.open("GuiDocument.xml")
        guidata = gf.read()
        gf.close()
        Handler = FreeCAD_xml_handler()
        xml.sax.parseString(guidata, Handler)
        guidata = Handler.guidata
        for key,properties in guidata.items():
            # open each diffusecolor files and retrieve values
            # first 4 bytes are the array length, then each group of 4 bytes is abgr
            if "DiffuseColor" in properties:
                #print ("opening:",guidata[key]["DiffuseColor"])
                df = zdoc.open(guidata[key]["DiffuseColor"])
                buf = df.read()
                #print (buf," length ",len(buf))
                df.close()
                cols = []
                for i in range(1,int(len(buf)/4)):
                    cols.append((buf[i*4+3],buf[i*4+2],buf[i*4+1],buf[i*4]))
                guidata[key]["DiffuseColor"] = cols
    zdoc.close()
    #print ("guidata:",guidata)

for key, value in guidata.items():
    names.append(key)
    diffuse_colors.append(value.get('DiffuseColor'))
```
</details>

storing here for the record

**Comment by yorikvanhavre** [comment]
i'd love to have somewhere an agnostic, universal, bi-directional FreeCAD-to-Blender-to-FreeCAD converter module, that could be used everywhere...

**Reply by zeffii** [reply]
in the meantime there's some experimenting to be done :)

<details>

```python
"""
>in FilePath FP
>in ObjNames T
out diffuse_colors v
"""
import zipfile
import xml.sax
import FreeCAD

class FreeCAD_xml_handler(xml.sax.ContentHandler):

    """A XML handler to process the FreeCAD GUI xml data"""

    # this creates a dictionary where each key is a FC object name,
    # and each value is a dictionary of property:value pairs

    def __init__(self):

        self.guidata = {}
        self.current = None
        self.properties = {}
        self.currentprop = None
        self.currentval = None

    # Call when an element starts

    def startElement(self, tag, attributes):

        if tag == "ViewProvider":
            self.current = attributes["name"]
        elif tag == "Property":
            name = attributes["name"]
            if name in ["Visibility","ShapeColor","Transparency","DiffuseColor"]:
                self.currentprop = name
        elif tag == "Bool":
            if attributes["value"] == "true":
                self.currentval = True
            else:
                self.currentval = False
        elif tag == "PropertyColor":
            c = int(attributes["value"])
            r = float((c>>24)&0xFF)/255.0
            g = float((c>>16)&0xFF)/255.0
            b = float((c>>8)&0xFF)/255.0
            self.currentval = (r,g,b)
        elif tag == "Integer":
            self.currentval = int(attributes["value"])
        elif tag == "Float":
            self.currentval = float(attributes["value"])
        elif tag == "ColorList":
            self.currentval = attributes["file"]

    # Call when an elements ends

    def endElement(self, tag):

        if tag == "ViewProvider":
            if self.current and self.properties:
                self.guidata[self.current] = self.properties
                self.current = None
                self.properties = {}
        elif tag == "Property":
            if self.currentprop and (self.currentval != None):
                self.properties[self.currentprop] = self.currentval
                self.currentprop = None
                self.currentval = None

filename = FilePath[0][0]
# check if we have a GUI document
guidata = {}
zdoc = zipfile.ZipFile(filename)
if zdoc:
    if "GuiDocument.xml" in zdoc.namelist():
        gf = zdoc.open("GuiDocument.xml")
        guidata = gf.read()
        gf.close()
        Handler = FreeCAD_xml_handler()
        xml.sax.parseString(guidata, Handler)
        guidata = Handler.guidata
        for key,properties in guidata.items():
            # open each diffusecolor files and retrieve values
            # first 4 bytes are the array length, then each group of 4 bytes is abgr
            if "DiffuseColor" in properties:
                #print ("opening:",guidata[key]["DiffuseColor"])
                df = zdoc.open(guidata[key]["DiffuseColor"])
                buf = df.read()
                #print (buf," length ",len(buf))
                df.close()
                cols = []
                for i in range(1,int(len(buf)/4)):
                    cols.append((buf[i*4+3],buf[i*4+2],buf[i*4+1],buf[i*4]))
                guidata[key]["DiffuseColor"] = cols
    zdoc.close()

for fullname, name, label in ObjNames:
    reference = guidata.get(name)
    if reference:
        color = reference.get("DiffuseColor")
        # some entries contain a multitude of colors.
        r,g,b,a = color[0]
        rgba = r/255, g/255, b/255, 1.0
        diffuse_colors.append([rgba])
```
</details>

![image](https://user-images.githubusercontent.com/619340/173230203-a8a3049a-9797-44e9-baf5-c6ae86d1a879.png)

**Reply by zeffii** [reply]
dissecting your importer, i'll be able to assign materials (diffuse colors) per face @yorikvanhavre  - i think, from what i can tell objects can have various regions with different diffuse colors

**Reply by yorikvanhavre** [reply]
Please dissect! :)
Yes. objects have a DiffuseColor property, which is a list that contains either: 1) one single rgba tuple (all object has the same color) or 2) one color per face. that list must then have the same length as the number of faces. face 1 has color 1, etc...

**Reply by zeffii** [reply]
cool, as expected :)

**Reply by zeffii** [reply]
@yorikvanhavre  would you be OK with me adding large swathes of your importer as a module into Sverchok ? -- and maybe slightly modifying it as progress is made?

**Comment by yorikvanhavre** [comment]
More than OK, I would be very glad ;)

**Reply by zeffii** [reply]
okidoki : )

**Reply by zeffii** [reply]
@yorikvanhavre  you'll recognize the FreeCAD_utils code, this change from a `verts list` to a `vdict = {(v.x, v.y, v.z): index, ......}` appears to result in a 6x+ increase in import speed for the geometry.

https://github.com/nortikin/sverchok/blob/acb8697a6b613ebe018aa2078d8f23c67d08c5e2/utils/modules/FreeCAD_utils.py

**Reply by zeffii** [reply]
i'm wondering if a dedicated dict class with rounding at 6..or 7 digits could be more optimal.. heck.. i'll try both :)

but first onwards to the materials/face-related stuff

**Reply by yorikvanhavre** [reply]
> @yorikvanhavre you'll recognize the FreeCAD_utils code, this change from a `verts list` to a `vdict = {(v.x, v.y, v.z): index, ......}` appears to result in a 6x+ increase in import speed for the geometry.

 Wow.. that's impressive! i'll do some testing too

**Reply by zeffii** [reply]
speedup to be around 6 to 10 times.. (not tried a dict wrapper class yet ) https://github.com/nortikin/sverchok/blob/fcstd_FCAD_dicts/utils/modules/FreeCAD_utils.py

**Comment by zeffii** [comment]
@yorikvanhavre  can you clear something up for me.  in the FreeCAD api.
```python
                F.closeDocument(doc.Name)   # 1.
                del doc                     # 2.
```
must the user explicitly do `1.`  or can we use `2.` , a 3rd option is to assume that (as is often with python) that as soon as the scope nolonger references a variable, that it is automatically discarded ?   or maybe if this is discussed somewhere that you know off the top of your head?


will an opened `doc` always have a `doc.Name` ?

**Reply by yorikvanhavre** [reply]
`#2` is not necessary. Indeed Python does it automatically. You generally use `del` only if you expressly want to (for example, to reduce memory use or if the variable could be accessed from somewhere else, and you don't want to). Note that if you are creating a temporary FreeCAD document, and that you save it with saveas(), and you don't need it anymore afterwards, you don't even need `#1` either.

Yes, FreeCAD documents always have a unmutable `Name`. Note that they can also have a `Label`, which is a user-modifiable name, that is usually a "friendlier" version of the name. In FreeCAD, the `Name` of documents and objects is unique, can only have ASCII characters, and is actually more like an ID...

**Comment by zeffii** [comment]
got the color stuff working.
[freecad_utils_test_2022_06_20_08_06.zip](https://github.com/nortikin/sverchok/files/8938746/freecad_utils_test_2022_06_20_08_06.zip)

![image](https://user-images.githubusercontent.com/619340/174555217-9a5f90bb-5968-4455-8c5d-52ab7fa548a6.png)

and using vertex colors :)

![image](https://user-images.githubusercontent.com/619340/174556929-1c542965-d68c-4a26-bce0-abd382cfbba0.png)

[freecad_utils_test_vcol_2022_06_20_08_17.zip](https://github.com/nortikin/sverchok/files/8938825/freecad_utils_test_vcol_2022_06_20_08_17.zip)

**Reply by zeffii** [reply]
configuring materials (setting them up especially) is not something Sverchok does yet,
